### PR TITLE
[Pallas TPU] Add lowering for `lax.erf_inv` 32 bit

### PR DESCRIPTION
**Pull request by ayaka14732**
Add lowering for `lax.erf_inv` 32 bit

### CODE REVIEW

**Comment by ayaka14732** [comment]
The implementation is based on https://github.com/openxla/xla/blob/a7a09d56c3599123f8148bbf3e44c9ebc04624b9/xla/mlir_hlo/mhlo/transforms/chlo_legalize_to_hlo/chlo_legalize_to_hlo.cc#L644-L802

File: jax/_src/pallas/mosaic/lowering.py [hunk 1]
```diff
@@ -2267,6 +2267,47 @@ def _shift_right_logical_lowering_rules(ctx: LoweringRuleContext, x, d):
...
+def _erf_inv_lowering_rule(ctx: LoweringRuleContext, x):
+  (x_aval,) = ctx.avals_in
+  if x_aval.dtype == jnp.float32:
+    return lower_fun(_erf_inv_32_helper, multiple_results=False)(ctx, x)
+  else:
+    raise NotImplementedError
+
+
+lowering_rules[lax.erf_inv_p] = _erf_inv_lowering_rule
+skip_mlir_conversions.add(lax.erf_inv_p)
```
Comment (sharadmv) [inline jax/_src/pallas/mosaic/lowering.py:2308]: You shouldn't need to skip MLIR conversions for this primitive (we don't need access to the literal values in the jaxpr).

File: tests/pallas/tpu/pallas_call_test.py [hunk 2]
```diff
@@ -2238,6 +2238,32 @@ def kernel(size_smem_ref, x_hbm_ref, _, o_hbm_ref, sem):
     np.testing.assert_array_equal(out, expected)
 
 
+class PallasCallOpsTest(PallasTPUTest):
```
Comment (sharadmv) [inline tests/pallas/tpu/pallas_call_test.py:2241]: Do you mind putting this in a new file?

**Review by sharadmv — CHANGES_REQUESTED** [review]

File: tests/pallas/tpu/pallas_call_test.py (hunk 2 above)
Comment (ayaka14732) [reply tests/pallas/tpu/pallas_call_test.py]: Done

File: jax/_src/pallas/mosaic/lowering.py (hunk 1 above)
Comment (ayaka14732) [reply jax/_src/pallas/mosaic/lowering.py]: Done

File: tests/pallas/tpu/pallas_call_test.py [hunk 3]
```diff
@@ -2238,6 +2238,32 @@ def kernel(size_smem_ref, x_hbm_ref, _, o_hbm_ref, sem):
...
+
+  def setUp(self):
```
[the remark this answers is not in the document]
Comment (ayaka14732) [reply tests/pallas/tpu/pallas_call_test.py]: Done

File: jax/_src/pallas/mosaic/lowering.py [hunk 4]
```diff
@@ -2267,6 +2267,47 @@ def _shift_right_logical_lowering_rules(ctx: LoweringRuleContext, x, d):
 skip_mlir_conversions.add(lax.shift_right_logical_p)
 
 
+# based on https://github.com/openxla/xla/blob/a7a09d56c3599123f8148bbf3e44c9ebc04624b9/xla/mlir_hlo/mhlo/transforms/chlo_legalize_to_hlo/chlo_legalize_to_hlo.cc#L644-L802
```
Comment (superbobry) [inline jax/_src/pallas/mosaic/lowering.py:2270]: I guess we can also create a new Erf op in Mosaic TPU, but this version has a benefit of working under both Mosaic and Triton.

File: jax/_src/pallas/mosaic/lowering.py [hunk 5]
```diff
@@ -2267,6 +2267,47 @@ def _shift_right_logical_lowering_rules(ctx: LoweringRuleContext, x, d):
...
+  w = jnp.where(lt, w - 2.5, jnp.sqrt(w) - 3.0)
+
+  p = jnp.where(lt, w_less_than_5_constants[0], w_greater_than_5_constants[0])
+  for i in range(1, k_degree):
+    c = jnp.where(lt, w_less_than_5_constants[i], w_greater_than_5_constants[i])
+    p = c + p * w
+
+  result = p * x
+  edge_cases = jnp.abs(x) == 1.0
+  result = jnp.where(edge_cases, jnp.inf * x, result)
```
Comment (superbobry) [inline jax/_src/pallas/mosaic/lowering.py:2296]: Nit: return directly.

File: jax/_src/pallas/mosaic/lowering.py [hunk 6]
```diff
@@ -2267,6 +2267,47 @@ def _shift_right_logical_lowering_rules(ctx: LoweringRuleContext, x, d):
...
+  lt = w < 5.0
+
+  w = jnp.where(lt, w - 2.5, jnp.sqrt(w) - 3.0)
+
+  p = jnp.where(lt, w_less_than_5_constants[0], w_greater_than_5_constants[0])
+  for i in range(1, k_degree):
+    c = jnp.where(lt, w_less_than_5_constants[i], w_greater_than_5_constants[i])
+    p = c + p * w
+
+  result = p * x
```
Comment (superbobry) [inline jax/_src/pallas/mosaic/lowering.py:2294]: Nit: I would personally inline both `p * x` and `edge_cases`.

File: jax/_src/pallas/mosaic/lowering.py [hunk 7]
```diff
@@ -2267,6 +2267,47 @@ def _shift_right_logical_lowering_rules(ctx: LoweringRuleContext, x, d):
...
+    -0.00417768164,  0.246640727,    1.50140941,
+  ]
+  w_greater_than_5_constants = [
+    -0.000200214257, 0.000100950558, 0.00134934322,
+    -0.00367342844,  0.00573950773,  -0.0076224613,
+    0.00943887047,   1.00167406,     2.83297682,
+  ]
+
+  w = -jnp.log1p(x * -x)
+  lt = w < 5.0
```
Comment (superbobry) [inline jax/_src/pallas/mosaic/lowering.py:2285]: w_lt_5?

File: jax/_src/pallas/mosaic/lowering.py [hunk 8]
```diff
@@ -2267,6 +2267,47 @@ def _shift_right_logical_lowering_rules(ctx: LoweringRuleContext, x, d):
 skip_mlir_conversions.add(lax.shift_right_logical_p)
 
 
+# based on https://github.com/openxla/xla/blob/a7a09d56c3599123f8148bbf3e44c9ebc04624b9/xla/mlir_hlo/mhlo/transforms/chlo_legalize_to_hlo/chlo_legalize_to_hlo.cc#L644-L802
+def _erf_inv_32_helper(x):
+  k_degree = 9
+  w_less_than_5_constants = [
```
Comment (superbobry) [inline jax/_src/pallas/mosaic/lowering.py:2273]: How about `w_lt_5_constants` and `w_ge_5_constants`?

File: jax/_src/pallas/mosaic/lowering.py [hunk 9]
```diff
@@ -2267,6 +2267,47 @@ def _shift_right_logical_lowering_rules(ctx: LoweringRuleContext, x, d):
...
+    -0.00367342844,  0.00573950773,  -0.0076224613,
+    0.00943887047,   1.00167406,     2.83297682,
+  ]
+
+  w = -jnp.log1p(x * -x)
+  lt = w < 5.0
+
+  w = jnp.where(lt, w - 2.5, jnp.sqrt(w) - 3.0)
+
+  p = jnp.where(lt, w_less_than_5_constants[0], w_greater_than_5_constants[0])
```
Comment (superbobry) [inline jax/_src/pallas/mosaic/lowering.py:2289]: Maybe fold this into the loop?

```
p = None
for lt_c, ge_c in zip(w_less_than_5_constants, ...):
  c = jnp.where(lt, lt_c, ge_c)
  if p is None:
    p = c
  else:
    p = c + p*w
```

You can also use `0` for `p`, but I guess we don't really want it appearing in the jaxpr, which is why I used `None` here.

File: tests/pallas/tpu/ops_test.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import jax.numpy as jnp
+from jax import lax
+from jax._src import test_util as jtu
+from jax.experimental import pallas as pl
+
+# Import mosaic for flag definitions
+from jax.experimental import mosaic as _  # noqa: F401
+
+
+class OpsTest(jtu.JaxTestCase):
```
Comment (superbobry) [inline tests/pallas/tpu/ops_test.py:31]: Shall we re-use pallas/ops_test.py?

File: jax/_src/pallas/mosaic/lowering.py (hunk 9 above)
Comment (ayaka14732) [reply jax/_src/pallas/mosaic/lowering.py:2289]: I prefer the current one, because there is no need to test `if p is None` every time (although it's the same on jaxpr level)

File: jax/_src/pallas/mosaic/lowering.py (hunk 8 above)
Comment (ayaka14732) [reply jax/_src/pallas/mosaic/lowering.py]: Sounds good. Done

File: jax/_src/pallas/mosaic/lowering.py (hunk 6 above)
Comment (ayaka14732) [reply jax/_src/pallas/mosaic/lowering.py]: Done

File: jax/_src/pallas/mosaic/lowering.py (hunk 7 above)
Comment (ayaka14732) [reply jax/_src/pallas/mosaic/lowering.py]: Done

File: tests/pallas/tpu/ops_test.py (hunk 10 above)
Comment (sharadmv) [reply tests/pallas/tpu/ops_test.py:31]: Ah yeah good call, didn't see we had that. Though will this pass on GPU?

Comment (superbobry) [reply tests/pallas/tpu/ops_test.py:31]: It should, since it's a lower_fun rule?

**Review by sharadmv — APPROVED** [review]
Thanks!

**Comment by ayaka14732** [comment]
```
________________ ERROR collecting tests/pallas/tpu/ops_test.py _________________
import file mismatch:
imported module 'ops_test' has this __file__ attribute:
  /home/runner/work/jax/jax/tests/pallas/ops_test.py
which is not the same as the test file we want to collect:
  /home/runner/work/jax/jax/tests/pallas/tpu/ops_test.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```

**Comment by ayaka14732** [comment]
Seems that the test cannot be named `ops_test` because there is already an `ops_test`. Since the ops tests in the `gpu` directory is called `gpu_ops_test`, I renamed the file in my PR to `tpu_ops_test`.

**Comment by superbobry** [comment]
Does the error still happen if you run via pytest (which is what JAX uses on the CI)?

**Comment by ayaka14732** [comment]
> Does the error still happen if you run via pytest (which is what JAX uses on the CI)?

The error happens on the CI and does not happen locally.

**Comment by superbobry** [comment]
It looks like this can be fixed by adding `__init__.py` files to the tests/pallas subdirectories. See https://github.com/pytest-dev/pytest/issues/3151#issuecomment-360493948.

Also, to repro the failure locally, do

```
python -m pytest --collect-only tests
```

**Comment by ayaka14732** [comment]
@superbobry Do you think the current naming `tpu_ops_test` is okay, or should I change back to `ops_test`?

**Comment by superbobry** [comment]
I'm fine either way. We can always rename in a follow up.
